File: T.E/SEM 5/IIS/pract/pract-3/code/DLS.java
class Tree{
    int val;
    Tree left;
    Tree right;
    Tree(int val){
        this.val = val;
        this.left = null;
        this.right = null;
    }

}

public class DLS{
    public static void main(String[] args) {
        Tree binaryTree = new Tree(1);
        binaryTree.left = new Tree(2);
        binaryTree.right = new Tree(3);

        binaryTree.left.left = new Tree(4);
        binaryTree.left.right = new Tree(5);

        binaryTree.right.left = new Tree(6);
        binaryTree.right.right = new Tree(7);

        int h = 2;
        int tar = 2;

        int x = searchDLS(binaryTree, h, tar) == 0? -1:searchDLS(binaryTree, h, tar);
        
        
        System.out.println(x);
    }

    private static int searchDLS(Tree bt, int h, int t){
        h--;
        if (h == -1) return 0;
        if(bt == null) return 0;

        System.out.println(bt.val);
        if(t == bt.val) return bt.val;
        int l = searchDLS(bt.left, h, t);
        if (l != 0) return l;
        int r = searchDLS(bt.right, h, t);
        if (r != 0) return r;
        return l==0?r:l;
    }
}

File: T.E/SEM 5/IIS/pract/IIS pract/My Code/DLS.py
class Tree:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None



def DLS(root, l, t):
    if(root == None):
        return
    
    if(l == 0):
        return
    
    if(root.data == t):
        # print(root.data)
        return root.data
    
    x1 = DLS(root.left, l-1, t)
    if(x1 == t):
        return x1
    x2 = DLS(root.right, l-1, t)
    if(x2 == t):
        return x2

tree = Tree('S')
tree.left = Tree('A')
tree.right = Tree('B')

tree.left.left = Tree('C')
tree.left.right = Tree('D')

tree.right.left = Tree('I')
tree.right.right = Tree('J')

tree.left.left.left = Tree('E')
tree.left.left.right = Tree('F')

tree.left.right.left = Tree('G')

tree.right.left.left = Tree('H')

target = 'J'
x = DLS(tree, 2, target)

if(x == None):
    print("Element not found")
else:
    print("Element found :", x)


# TC: O(b^l)
# SC: S(b*l)